src/kudu/consensus/consensus_queue.cc

```diff
@@ -1332,12 +1333,14 @@ int64_t PeerMessageQueue::ComputeNewWatermarkStaticMode(int64_t* watermark) {
       queue_state_.active_config->voter_distribution().begin(),
       queue_state_.active_config->voter_distribution().end());
 
-  // Compute number of voters in each region in the active config.
-  // As voter distribution provided in topology config can lag,
-  // we need to take into account the active voters as well due to
-  // membership changes.
-  AdjustVoterDistributionWithCurrentVoters(
-      *(queue_state_.active_config), &voter_distribution);
+  if (adjust_voter_distribution_) {
```

Comment:
You probably do not want this change and only want `AdjustVoterDistributionWithCurrentVoters(...)` to not be called from leader election module (the one called from `FlexibleVoteCounter::FetchTopologyInfo()`). This one here is counting votes for regular raft operations (trxs, rotates, no-ops). If so, then you can remove all changes in consensus_queue.

Reply:
Let me see if this is required. However, adjusting voter dist on in sub set of code path seems a little weird. If this is not required I'll probably change the var name to adjust_voter_distribution_during_election

Reply:
This is likely not going to matter because of the constrained space in which voter distributions will be bypassed.
This piece of code is used to determine commit of trxs and No-Op will be the only trx which will be committed on new LEADER. In future changes to Flexi-Raft, we could have quorums which AND conditions and in such situations the commit could need the lower voter distribution on the previous region as well. Although in Single region dynamic this would likely not be the case as the commit will be in-region of new leader.

So lets keep this as is and we can revisit this later.

Reply:
IMO, this should be removed.
1. Offline checks picks the longest log as the new candidate
2. New candidate wins the election
3. As part of writing NoOp, the new leader commits all trxs from previous term. If this commit does not follow data commit rule and avoids adjusting voter distribution, then all the trxs committed as part of NoOp may not reach majority.

So, even though the unsafe-config-change did not give any warnings about data drift/data loss, a subsequent election might end up losing data or leading to data drifts. So, IMO, any 'commits' (even NoOp) should continue to honor data commit rules even during unsafe-config-change

Reply:
I agree with you for simple dynamic quorums but there can be some static configuration such that the broken region is part of the data quorum. In such cases we won't be able to commit the NoOp event. UCC will also do some pre-checks and quiescing where it'll disallow config changes (add/remove members) to reduce the likely hood of hitting such edge cases. Overall it might be better for UCC to barge thru and shift the leader even if it ends up violating raft properties for a while rather than it getting stuck without resolving a broken ring. Thoughts?

Reply:
@anirbanr-fb, @yashtc and I discussed this issue and think that it's better to handle this issue by making UCC block member addition and by making it commit a transaction after voter dist adjustment is re-enabled so that the majority of quorum from committed config is consistent. While this will not completely solve the issue it reduces the probability of dataloss if the mini majority (the majority only based on voter dist) that was used some commit trxs (including NoOp) goes down. Ultimately, it's more advantageous for UCC to commit NoOp than leaving the ring in a limbo state if NoOp cannot be committed.